src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionEvent.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.event;
+
+import java.util.Map;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.Level;
+import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.ICancellableEvent;
+import net.neoforged.fml.LogicalSide;
+import net.neoforged.fml.event.IModBusEvent;
+import net.neoforged.neoforge.client.DimensionTransitionScreen;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * <p>Event for registering screen effects when transitioning across dimensions.</p>
+ *
+ * <p>This event is not {@linkplain ICancellableEvent cancellable}, and does not {@linkplain HasResult have a result}.
+ *
+ * <p>This event is fired on the mod-specific event bus, only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterDimensionTransitionEvent extends Event implements IModBusEvent {
+    private final Map<ResourceKey<Level>, DimensionTransitionScreen> toEffects;
+    private final Map<ResourceKey<Level>, DimensionTransitionScreen> fromEffects;
+
+    @ApiStatus.Internal
+    public RegisterDimensionTransitionEvent(Map<ResourceKey<Level>, DimensionTransitionScreen> toEffects, Map<ResourceKey<Level>, DimensionTransitionScreen> fromEffects) {
+        this.toEffects = toEffects;
+        this.fromEffects = fromEffects;
+    }
+
+    /**
+     * Registers a dimension transition when traveling to a dimension.
+     */
+    public void registerTo(ResourceKey<Level> dimension, DimensionTransitionScreen screen) {
```

Comment:
Are the names clear enough? We could perhaps name these to `outgoing`/`incoming`. (Just a thought.)